src/QsCompiler/CompilationManager/DataStructures.cs

```diff
@@ -41,69 +50,374 @@ internal class CodeLine
+                var isDocComment = string.IsNullOrWhiteSpace(this.WithoutEnding) && (commentStr.Length - commentStr.TrimStart('/').Length == 3);
+                this.EndOfLineComment = isDocComment ? null : commentStr.Substring(2);
+            }
+            // else there is no comment
+            else
+            {
+                this.WithoutEnding = text.Substring(0, lineLength);
+                this.EndOfLineComment = null;
+            }
+
+            this.StringDelimiters = delimiterBuilder.ToImmutable();
+            this.ErrorDelimiterPositions = errorDelimiterBuilder.ToImmutable();
+            this.Indentation = indentation;
+            this.ExcessBracketPositions = excessBrackets.ToImmutableArray();
+            ScopeTracking.VerifyStringDelimiters(text, this.StringDelimiters);
+            ScopeTracking.VerifyExcessBracketPositions(this, excessBrackets);
         }
 
-        internal CodeLine SetStringDelimiters(IEnumerable<int> newStringDelimiters)
+        /// <summary>
+        /// Computes the location of the string delimiters within a given text.
+        /// </summary>
+        private static IEnumerable<int> ComputeStringDelimiters(string text, ref StringContext stringContext, out int commentIndex, out IEnumerable<int> errorDelimiters)
         {
-            return new CodeLine(this.Text, newStringDelimiters, this.WithoutEnding.Length, this.Indentation, this.ExcessBracketPositions);
+            var delimiterBuilder = ImmutableArray.CreateBuilder<int>();
+            var errorDelimiterBuilder = ImmutableArray.CreateBuilder<int>();
+
+            if (stringContext >= StringContext.OpenString)
+            {
+                delimiterBuilder.Add(-1);
+            }
+
+            commentIndex = -1;
+            var stringLength = text.Length;
+            while (text != string.Empty)
+            {
+                // Check for a comment start if we are outside a string
+                commentIndex = stringContext >= StringContext.OpenString ? -1 : text.IndexOf("//");
+
+                var index = -1;
+                switch (stringContext)
+                {
+                    case StringContext.NoOpenString:
+                        // Find the next "
+                        index = text.IndexOf('"');
+                        break;
+                    case StringContext.OpenInterpolatedArgument:
+                        // Find the next " or }
+                        index = text.IndexOfAny(new[] { '"', '}' });
+                        break;
+                    case StringContext.OpenInterpolatedString:
+                        // Find the next " or {, neither or which being preceded by \
+                        index = text.IndexOfAny(new[] { '"', '{' });
+                        while (index > 0 && text[index - 1] == '\\')
+                        {
+                            var next = text.Substring(index + 1).IndexOfAny(new[] { '"', '{' });
```

Comment:
Instead of allocating these arrays dynamically, you can allocated it once (static modifier) and use it in several places.

Reply:
Good idea, done